community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/command/IndexBatchTransactionApplier.java

```diff
@@ -93,11 +108,126 @@ public void close() throws Exception
             labelScanStoreSync.apply( new LabelUpdateWork( labelUpdates ) );
         }
 
-        if ( affectedIndexes != null )
+        // Apply all the index updates within this whole batch of transactions.
+        applyIndexUpdates();
+    }
+
+    /**
+     * Made as an internal non-static class here since the batch applier has so much interaction with
+     * the transaction applier such that keeping them apart would incur too much data structures and interfaces
+     * purely for communicating between the two to make the code hard to read.
+     */
+    private class SingleTransactionApplier extends TransactionApplier.Adapter
+    {
+        private final NodeStore nodeStore;
+        private final PropertyStore propertyStore;
+        private final PropertyLoader propertyLoader;
+        private final TransactionApplicationMode mode;
+        private final NodePropertyCommandsExtractor indexUpdatesExtractor = new NodePropertyCommandsExtractor();
+
+        public SingleTransactionApplier( NodeStore nodeStore, PropertyStore propertyStore,
+                PropertyLoader propertyLoader, TransactionApplicationMode mode )
+        {
+            this.nodeStore = nodeStore;
+            this.propertyStore = propertyStore;
+            this.propertyLoader = propertyLoader;
+            this.mode = mode;
+        }
+
+        @Override
+        public void close() throws Exception
         {
-            // Since we have written changes to indexes w/o refreshing readers, then do so now
-            // at this point where all changes in this whole batch of transactions have been applied.
-            indexingService.flushAll( affectedIndexes );
+            if ( indexUpdatesExtractor.containsAnyNodeOrPropertyUpdate() )
+            {
+                // Queue the index updates. When index updates from all transactions in this batch have been accumulated
+                // we'll feed them to the index updates work sync at the end of the batch
+                indexUpdates = mode == TransactionApplicationMode.RECOVERY ?
```

Comment:
i was thinking about something like this:
`indexUpdates = createIndexUpdates();`
and factory method looks like
`private ValidatedIndexUpdates createIndexUpdates() {
return TransactionApplicationMode.RECOVERY == mode ? new RecoveryIndexUpdates() : new OnlineIndexUpdates( nodeStore, propertyStore, propertyLoader );
}`
just because it so much easier to read